src/fireedge/src/modules/utils/labels.js

```diff
@@ -0,0 +1,226 @@
+  } catch {
+    try {
+      return Object.fromEntries(
+        labelString.split(',').map((lbl) => [`$${lbl}`, {}])
+      )
+    } catch {
+      return labelString
+    }
+  }
+}
+
+/**
+ * @param {object} params - Params
+ * @returns {object} - Labels base state
+ */
+export const buildLabelTreeState = (params) => {
+  const {
+    user: { ID: uId },
+    isOneAdmin,
+    groups,
+    labels,
+    selectedRows,
+    resourceType,
+  } = params
+
+  const selectedRowIds = new Set(
+    selectedRows?.map(({ original }) => original?.ID)
+  )
+
+  const adminGroupNames = new Set(
+    groups
+      ?.filter((group) => [].concat(group?.ADMINS?.ID).includes(uId))
+      ?.map((group) => group?.NAME)
+  )
+
+  const createMetadata = ({ expanded = false, isEditable = false }) => ({
+    children: {},
+    expanded,
+    selected: false,
+    indeterminate: false,
+    ids: [],
+    isEditable,
+  })
+
+  let treeRoot = null
+
+  const walk = (node, isRoot = true, nodePath = [], currentResult = {}) => {
+    if (isRoot) {
+      treeRoot = currentResult
+    }
+    for (const [nodeKey, currentNode] of Object.entries(node)) {
+      const isPlainObject =
+        typeof currentNode === 'object' &&
+        currentNode !== null &&
+        !Array.isArray(currentNode)
+
+      if (!isPlainObject) {
+        currentResult[nodeKey] = currentNode // Assign and skip, this is a resource node
+        continue
+      }
+
+      const path = [...nodePath, nodeKey]
+
+      const isUserEditable = path?.[0] === 'user'
+      const isGroupEditable =
+        isOneAdmin || (path?.[0] === 'group' && adminGroupNames?.has(path?.[1]))
+
+      const isEditable = isUserEditable || isGroupEditable
+
+      // Always expand root nodes or if we have no selected rows
+      const expanded = isRoot || selectedRowIds?.size <= 0
+
+      const nodeMeta = createMetadata({ expanded, isEditable })
+      currentResult[nodeKey] = nodeMeta
+
+      const resourceIdMap = Object.fromEntries(
+        Object.entries(currentNode).filter(
+          ([k, v]) =>
+            !k.startsWith('$') &&
+            Object.values(RESOURCE_NAMES).includes(k) &&
+            Array.isArray(v)
+        )
+      )
+
+      const matchingResourceIds =
+        resourceIdMap?.[resourceType]?.filter((id) =>
+          selectedRowIds?.has(id)
+        ) ?? []
+
+      nodeMeta.ids = resourceIdMap
+
+      if (matchingResourceIds?.length > 0) {
+        nodeMeta.selected = true
+        nodeMeta.expanded = true
+
+        // Expand ancestors if current node is expanded
+        path.slice(0, -1).reduce((acc, segment) => {
+          const next = acc?.[segment]
+          if (next) next.expanded = true
+
+          return next?.children
+        }, treeRoot)
+
+        if (matchingResourceIds.length !== selectedRowIds.size) {
+          nodeMeta.indeterminate = true
+        }
+      }
+
+      Object.values(RESOURCE_NAMES).forEach((r) => delete currentNode[r])
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

We detected that you are using `delete` to remove items from an array. If we use `delete`, it will delete the object property, but will not reindex the array or update its length. It leaves `undefined` holes in the array. Therefore, we recommend you to use `splice` to remove items from an array.

[Learn more](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/delete)